src/library_int53.js

```diff
@@ -0,0 +1,103 @@
+#endif
+  },
+
+  // Same as writeI53ToI64, but if the double precision number does not fit within the
+  // 64-bit number, the number is clamped to range [-2^63, 2^63-1].
+  $writeI53ToI64Clamped: function(ptr, num) {
+    if (num > 0x7FFFFFFFFFFFFFFF) {
+      HEAPU32[ptr>>2] = 0xFFFFFFFF;
+      HEAPU32[ptr+4>>2] = 0x7FFFFFFF;
+    } else if (num < -0x8000000000000000) {
+      HEAPU32[ptr>>2] = 0;
+      HEAPU32[ptr+4>>2] = 0x80000000;
+    } else {
+      HEAPU32[ptr>>2] = num;
+      HEAPU32[ptr+4>>2] = (num - HEAPU32[ptr>>2])/4294967296;
+    }
+  },
+
+  // Like writeI53ToI64, but throws if the passed number is out of range of int64.
+  $writeI53ToI64Signaling: function(ptr, num) {
+    if (num > 0x7FFFFFFFFFFFFFFF || num < -0x8000000000000000) {
+#if ASSERTIONS
+      throw 'RangeError in writeI53ToI64Signaling(): input value ' + num + ' is out of range of int64';
+#else
+      throw 'RangeError:'+num;
+#endif
+    }
+    HEAPU32[ptr>>2] = num;
+    HEAPU32[ptr+4>>2] = (num - HEAPU32[ptr>>2])/4294967296;
+  },
+
+  // Uint64 variant of writeI53ToI64Clamped. Writes the Number to a Uint64 variable on
+  // the heap, clamping out of range values to range [0, 2^64-1].
+  $writeI53ToU64Clamped: function(ptr, num) {
+    if (num > 0xFFFFFFFFFFFFFFFF) HEAPU32[ptr>>2] = HEAPU32[ptr+4>>2] = 0xFFFFFFFF;
+    else if (num < 0) HEAPU32[ptr>>2] = HEAPU32[ptr+4>>2] = 0;
+    else {
+      HEAPU32[ptr>>2] = num;
+      HEAPU32[ptr+4>>2] = (num - HEAPU32[ptr>>2])/4294967296;
+    }
+  },
+
+  // Like writeI53ToI64, but throws if the passed number is out of range of uint64.
+  $writeI53ToU64Signaling: function(ptr, num) {
+    if (num < 0 || num > 0xFFFFFFFFFFFFFFFF) {
+#if ASSERTIONS
+      throw 'RangeError in writeI53ToU64Signaling(): input value ' + num + ' is out of range of uint64';
+#else
+      throw 'RangeError:'+num;
+#endif
+    }
+    HEAPU32[ptr>>2] = num;
+    HEAPU32[ptr+4>>2] = (num - HEAPU32[ptr>>2])/4294967296;
+  },
+
+  // Reads a 64-bit signed integer from the WebAssembly heap and
+  // converts it to a JavaScript Number, which can represent 53 integer bits precisely.
+  // TODO: Add $readI53FromI64Signaling() variant.
+  $readI53FromI64: function(ptr) {
+    return HEAPU32[ptr>>2] + HEAP32[ptr+4>>2] * 4294967296;
+  },
+
+  // Reads a 64-bit unsigned integer from the WebAssembly heap and
+  // converts it to a JavaScript Number, which can represent 53 integer bits precisely.
+  // TODO: Add $readI53FromU64Signaling() variant.
+  $readI53FromU64: function(ptr) {
+    return HEAPU32[ptr>>2] + HEAPU32[ptr+4>>2] * 4294967296;
+  },
+
+  // Converts the given signed 32-bit low-high pair to a JavaScript Number that can
+  // represent 53 bits of precision.
+  // TODO: Add $convertI32PairToI53Signaling() variant.
+  $convertI32PairToI53: function(lo, hi) {
+    lo = lo >>> 0;
```

Comment:
to be safe why not do ` | 0` on hi? (like the old code used to)

Reply:
Hmm, can you think of a scenario where this would be important? I wonder perhaps instead an ASSERTIONS check for `assert(hi === hi|0);` would be better?

Reply:
Perhaps someone using the API happens to send in `convertI32PairToI53(x >>> 0, y >>> 0)`, either directly like I wrote there, or maybe it's output from another API where it arrived as unsigned. I think it's better to cast to be sure. However I agree for code size the assertion would be smaller, and this is an internal API, so I'm ok with either.